Add new CRT glshaders

dreamer opened the pull request:
```    
    Automatically adapt initial window size to available screen size to
    prevent problems with going in/out fullscreen when using advanced shader
    with windowresolution = original.
    
    Both shaders were copied from Tyrell's collection of well-known shaders
    adapted to DOSBox SVN format.
    
    Shader 'crt-easymode-flat' is recommended for game resolutions over
    640x480. Shader program was copied from crt-easymode.tweaked.glsl.
    
    Shader 'crt-fakelottes-flat' looks good for game resolutions up to
    640x480. Shader program was copied from fakelottes.tweaked.glsl.
```

kcgen commented on src/dosbox.cpp:
```diff
@@ -507,12 +507,15 @@ void DOSBOX_Init(void) {
-	                  "scan3x, tv2x, tv3x, sharp.");
+	pstring = secprop->Add_path("glshader", always, "sharp");
+	pstring->Set_help(
+	        "Path to GLSL shader source to use with OpenGL output\n"
+	        "(\"none\" to disable).  Can be either an absolute path,\n"
```
Minor suggestion: _"Either \"none\" or a GLSL shader name. Only works with OpenGL output. The shader can be an absolute path, a file in ... "_

dreamer replied:
Done :)

kcgen commented on src/dosbox.cpp:
```diff
@@ -507,12 +507,15 @@ void DOSBOX_Init(void) {
+	        "a file in the \"glshaders\" subdirectory of the DOSBox\n"
+	        "configuration directory, or one of the built-in shaders:\n"
+	        "advinterp2x, advinterp3x, advmame2x, advmame3x,\n"
+	        "crt-easymode-flat, crt-fakelottes-flat, rgb2x, rgb3x,\n"
```
Did some testing with these, and feel we need to add some explanation/recommendations for the CRT shaders. Perhaps beneath the list? 

- CRT shaders simulate an analog CRT on 1080p and 4k LCD-panels
- `crt-fakelottes-flat` and `output = opengl` should be used full screen on 1080p displays
- `crt-easymode-flat` and `output = openglnb or pp` should be used on 4k displays
- The effect of both CRT shaders diminishes at game resolutions above 640x480

dreamer replied:
They both work just fine on my 1366x768, and I think on 720p screen as well - of course, I bet they look much better on high-resolution screens (especially in text modes), but they look *fine* on my old laptop :)

They start to look crappy at 800x600 or smaller resolution - the code I added for autodetecting window size for `windowresolution=default` should prevent that from happening unless user wants to :)

I see no reason for confusing users and recommending different output modes per shader - I see no difference, example:

`opengl`:
![opengl-Screenshot from 2020-12-02 16-26-30](https://user-images.githubusercontent.com/3967/100893473-b1b6ad80-34bb-11eb-8e2d-2ca89f604e39.png)

`openglnb`:
![opnglnb-Screenshot from 2020-12-02 16-27-07](https://user-images.githubusercontent.com/3967/100893519-bbd8ac00-34bb-11eb-9a9f-aa778db15e95.png)

`openglpp`:
![openglpp-Screenshot from 2020-12-02 16-28-08](https://user-images.githubusercontent.com/3967/100893619-d90d7a80-34bb-11eb-98ea-c6487367f78d.png)

kcgen replied:
Oh - that's good news then!

All of my screenshots in Wayland are broken (fullscreen mode the surface is empty; window-mode the images are super-dar).. so I wasn't able to to do post-gameplay comparisons.

Let's drop those two suggestions.

kcgen replied:
Perhaps we should still mention: `CRT shaders simulate an analog CRT on LCD-panels`?

kcgen commented on src/gui/render.cpp:
```diff
@@ -615,26 +618,35 @@ void RENDER_SetForceUpdate(bool f) {
-static bool RENDER_GetShader(std::string& shader_path, char *old_src) {
+static bool RENDER_GetShader(std::string &shader_path, char *old_src)
+{
+	const std::unordered_map<std::string, const char *> builtin_shaders = {
```
Nice!

kcgen commented on src/gui/render.cpp:
```diff
@@ -615,26 +618,35 @@ void RENDER_SetForceUpdate(bool f) {
+	        {"advmame3x", advmame3x_glsl},
+	        {"crt-easymode-flat", crt_easymode_tweaked_glsl},
+	        {"crt-fakelottes-flat", crt_fakelottes_tweaked_glsl},
+	        {"rgb2x", rgb2x_glsl},
+	        {"rgb3x", rgb3x_glsl},
+	        {"scan2x", scan2x_glsl},
+	        {"scan3x", scan3x_glsl},
+	        {"sharp", sharp_glsl},
+	        {"tv2x", tv2x_glsl},
+	        {"tv3x", tv3x_glsl},
```
Almost time we can retire `scalar = <software implementation>` for good!

dreamer replied:
There's a biiig problem with retiring them :( Scalars are exposed via our CLI interface, and consequently they are being used by GUI frontends this way (e.g. Lutris). Also, glshaders for some of those legacy options seem to be sub-par.

kcgen replied:
Yes; we need to never break the user experience or external tools. 

Perhaps `scalar` (and command line argument) could map over to the equivalent `glshader`, and even try opengl output. If that doesn't work we could fallback to texture mode, and then recommend the user get opengl working on their device.

kcgen replied:
> glshaders for some of those legacy options seem to be sub-par.

That's a more serious problem.. I suppose they're just as bad even with openglnb or pp?

dreamer replied:
Hmm, I don't remember if I tested those older glshaders with openglnb… but if it matters, we should make it switch automatically, so user doesn't need to.

kcgen commented on src/gui/render.cpp:
```diff
@@ -615,26 +618,35 @@ void RENDER_SetForceUpdate(bool f) {
 	std::stringstream buf;
 	std::ifstream fshader(shader_path.c_str(), std::ios_base::binary);
-	if (!fshader.is_open()) fshader.open((shader_path + ".glsl").c_str(), std::ios_base::binary);
+	if (!fshader.is_open())
+		fshader.open(shader_path + ".glsl", std::ios_base::binary);
 	if (fshader.is_open()) {
 		buf << fshader.rdbuf();
 		fshader.close();
+	} else if (builtin_shaders.count(shader_path) > 0) {
+		buf << builtin_shaders.at(shader_path);
```
When I was testing shaders, I kept checking the console to see confirmation if the shader was loaded, and (after taking screenshots), I looked back to see which shader I was using.. but we're not logging anything.

Would be nice to add a one-liner output on success and the name of the active shader.

dreamer replied:
I believe a log is printed when the name is incorrect / shader file cannot be found. I'll add info log as well :)

kcgen commented on src/gui/render.cpp:
```diff
@@ -615,26 +618,35 @@ void RENDER_SetForceUpdate(bool f) {
-	else if (shader_path == "scan3x")      buf << scan3x_glsl;
-	else if (shader_path == "tv2x")        buf << tv2x_glsl;
-	else if (shader_path == "tv3x")        buf << tv3x_glsl;
-	else if (shader_path == "sharp")       buf << sharp_glsl;
```
Same with a more helpful log message on failure.

Currently it only produces the unhelpful `"WARNING: Couldn't copy shader source"` message; and that's not even under all conditions.

dreamer replied:
Uhh, I think it's out of scope… This algorithm for parametrizing shaders needs to be cleaned up anyway...

kcgen commented on src/gui/render.cpp:
```diff
@@ -615,26 +618,35 @@ void RENDER_SetForceUpdate(bool f) {
+		buf << builtin_shaders.at(shader_path);
 	}
-	else if (shader_path == "advinterp2x") buf << advinterp2x_glsl;
-	else if (shader_path == "advinterp3x") buf << advinterp3x_glsl;
```
Such a nice use of the map to eliminate this hardcoded boolean-centipede!

kcgen commented on src/gui/sdlmain.cpp:
```diff
@@ -1924,23 +1936,61 @@ static bool DetectResizableWindow()
 #endif // C_OPENGL
 }
 
+static bool detect_shader_fixed_size()
+{
+#if C_OPENGL
+	if (sdl.desktop.want_type != SCREEN_OPENGL)
+		return false;
+
+	const std::string sname = get_glshader_value();
+	return (sname == "crt-fakelottes-flat" || sname == "crt-easymode-flat");
+#else
+	return false;
+#endif // C_OPENGL
+}
```
This has a bit of a code-smell to it, given we're having to hard-code properties about items defined somewhere else in the code.

It would be nice for these properties to be stapled to the shaders, in the underordered map.
Maybe as a `pair` with these properties defined as an `enum`; or even make the shader a struct.
**shader:**
- name (const std::string)
- code (const char *)
- fixed_size (const bool)

This would also give us the ability take this property in dynamically for user-defined shader files (but in a subsequent PR).  Because right now, a user providing a fixed-size shader via filename/path would need the same name hardcoded here..).

dreamer replied:
Hmm, this should probably be a new API function, e.g. `RENDER_GetSelectedShaderFlags`…

Let's do this after 0.76.0. - I added a sticky note with this task to 0.77.0 release.

kcgen commented on src/gui/sdlmain.cpp:
```diff
@@ -1924,23 +1936,61 @@ static bool DetectResizableWindow()
+	SDL_Rect bounds;
+	SDL_GetDisplayBounds(sdl.display_number, &bounds);
+	constexpr int resolutions[][2] = {
+	        {1600, 1200},
+	        {1280, 960},
+	        {1024, 768},
+	        {800, 600},
```
Are these DOS-side resolutions?

dreamer replied:
No, these are user screen size resolutions - e.g. if user's desktop area is big enough to use window size 1600x1200, then we pick this as the default window size (only when `windowresolution = default` and specific shaders are selected).

kcgen replied:
Thanks for the explanation (could that be added as a comment?)
These handle the the 4k range:
```
{2560, 1920}
{2400, 1800}
{2048, 1536}
{1920, 1440}
```

dreamer replied:
Hmm, I'll add these resolutions but commented-out for now. And I commented-out 1600x1200 for now as well.

We want this to work, but I have no idea how this will behave for users who have pixel-doubling / display scaling enabled on OS level. This will require *extensive* testing on Windows, Linux, and macOS :( 

Resolutions 1280x960 and below are fine - even if OS will double the window size, users of 4K screens will end up with 2560x1920 window which will still fit on screen.

We can't depend on it working correctly now, because SDL2 had some issues with high density displays IIRC and we have non-SDL2 hack for pixel doubling on Windows inherited from our upstream.

kcgen replied:
Good catch with the line/pixel doubling. I think I've seen this in some modes, and it blows out the image so you only see a section of the DOS display. This is a nice safe subset for now.

kcgen commented on src/gui/sdlmain.cpp:
```diff
@@ -1924,23 +1936,61 @@ static bool DetectResizableWindow()
+	        {800, 600},
+	};
+	for (const auto &wh : resolutions) {
+		if (bounds.w > wh[0] && bounds.h > wh[1])
+			return std::make_tuple(wh[0], wh[1]);
+	}
+	return std::make_tuple(640, 480);
```
It's unfortunate SDL doesn't provide a simple `SDL_dims` object with width and height members. It only provide `SDL_rect`.
This code could just be cleaner `dim.w`, `dim.h`, etc.. and `return Dim(w, h)` instead of `make_tuple(wh[0], wh[1])`.. which feels cumbersome.

dreamer replied:
With SDL_Point, being a simple POD struct `return {640, 480};` should work - let's see if it passes CI :)

kcgen replied:
Working nicely :-)

kcgen commented on src/gui/sdlmain.cpp:
```diff
@@ -1924,23 +1936,61 @@ static bool DetectResizableWindow()
+
+		if (!sdl.desktop.want_resizable_window &&
+		    detect_shader_fixed_size()) {
+			const auto wh = detect_window_size();
+			sdl.desktop.window.use_original_size = false;
+			sdl.desktop.window.width = std::get<0>(wh);
+			sdl.desktop.window.height = std::get<1>(wh);
```
Using `SDL_rect`  (or a `Dims` object) would tidy this up too.

dreamer replied:
SDL_Rect has 4 members though, can't be constructed out of 2 integers :( There's SDL_Point though.

kcgen replied:
> SDL_Rect has 4 members though 

Indeed.. a simple Dims object would be perfect.

dreamer replied:
I used `SDL_Point`; it has fields `x` and `y`, but it's more readable than a tuple anyway.

kcgen replied:
Looks like a nice compromise! x and y are perfectly intuitive and still represent the axis nicely. Maybe we can use `Point` for more refactoring (instead of carrying around separate sizes or dimensions).

kcgen reviewed (changes requested):
Some suggestions and thoughts.

dreamer commented:
I implemented some suggestions, but not all - and postponed some for after the release.

Added a tiny change: the default value for `glshader` is now `default` (which redirects to `sharp`) - in case we'll want to change it in the future.

kcgen reviewed (approved):
All nice improvements @dreamer ! Thanks.

dreamer commented:
@kcgen Thanks for the review :)

We are only missing some final testing and polish / proof-reading of release notes.
